// @shapediver/viewer.settings/dist/versions/v4_1/Defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Defaults = void 0;
const Defaults = () => {
    return {
        build_date: '',
        build_version: '',
        settings_version: '4.1',
        ar: {
            enable: true,
            autoScaling: true
        },
        camera: {
            cameraId: '',
            cameras: {},
        },
        environment: {
            clearAlpha: 1.0,
            clearColor: '#ffffff',
            map: 'photo_studio',
            mapAsBackground: false,
            mapResolution: '1024',
            rotation: { x: 0, y: 0, z: 0, w: 1 },
            intensity: 1,
            blurriness: 0
        },
        environmentGeometry: {
            gridColor: '#44444426',
            gridVisibility: false,
            groundPlaneColor: '#636363ff',
            groundPlaneVisibility: true,
            groundPlaneShadowColor: '#d3d3d3ff',
            groundPlaneShadowVisibility: false,
        },
        general: {
            transformation: {
                scale: { x: 1, y: 1, z: 1 },
                translation: { x: 0, y: 0, z: 0 },
                rotation: { x: 0, y: 0, z: 0 }
            },
            blurWhenBusy: true,
            commitSettings: false,
            commitParameters: false,
            pointSize: 1.0,
            showMessages: true,
            defaultMaterialColor: '#199b9bff'
        },
        light: {
            lightSceneId: '',
            lightScenes: {},
        },
        postprocessing: {
            antiAliasingTechnique: "smaa",
            antiAliasingTechniqueMobile: "fxaa",
            enablePostProcessingOnMobile: true,
            ssaaSampleLevel: 2,
            effects: [
                {
                    properties: {
                        resolutionScale: 1,
                        spp: 8,
                        distance: 1,
                        distanceIntensity: 1,
                        intensity: 2.5,
                        color: '#000000',
                        iterations: 1,
                        radius: 15,
                        rings: 4,
                        lumaPhi: 10,
                        depthPhi: 2,
                        normalPhi: 3.25,
                        samples: 16,
                    },
                    type: "ssao"
                }
            ]
        },
        rendering: {
            automaticColorAdjustment: true,
            beautyRenderDelay: 50,
            beautyRenderBlendingDuration: 1500,
            lights: true,
            outputEncoding: 'srgb',
            physicallyCorrectLights: true,
            shadows: true,
            softShadows: true,
            textureEncoding: 'srgb',
            toneMapping: 'none',
            toneMappingExposure: 1,
        },
        session: {},
    };
};
exports.Defaults = Defaults;
