// @shapediver/api.geometry-api-dto-v2/src/commons/SdCommonsParameter.ts
/**
 * @swagger {components.schemas} CommonsParameterBasic {
     "description": "Definition of a basic ShapeDiver parameter.",
     "oneOf": [
       { "type": "string" },
       { "type": "number" },
       { "type": "boolean" }
     ]
   }
 * @example
 * "0014bb77-e57c-4be3-92eb-24b59c41ee43" // File
 * "0x11223344" // Color
 * "0001-01-01T23:55:00" // Time
 */
/** Definition of a basic ShapeDiver parameter. */
export type ShapeDiverCommonsParameterBasic = string | number | boolean;

/**
 * @swagger {components.schemas} CommonsParameterSType {
     "description": "Definition of the value to use for s-type parameters.",
     "type": "object"
   }
 */
/** Definition of the value to use for s-type parameters. */
export interface ShapeDiverCommonsParameterSType {
    /**
     * @swagger {components.schemas.CommonsParameterSType.properties} value {
         "description": "Optional embedded value. If this is set the asset is ignored.",
         "type": "string"
       }
     */
    /** Optional embedded value. If this is set the asset is ignored. */
    value?: string;

    /**
     * @swagger {components.schemas.CommonsParameterSType.properties} asset {
         "description": "Reference to the asset to be used.",
         "type": "object"
       }
     */
    /** Reference to the asset to be used. */
    asset?: {
        /**
         * @swagger {components.schemas.CommonsParameterSType.properties.asset.properties} id {
              "description": "String ID of the asset.",
              "type": "string"
            }
         * @swagger {components.schemas.CommonsParameterSType.properties.asset} required [ "id" ]
         */
        /** String ID of the asset. */
        id: string;

        /**
         * @swagger {components.schemas.CommonsParameterSType.properties.asset.properties} chunk {
             "description": "Describing which chunk to use, instead of the default one, which would get chosen based on parameter ID and name.",
             "type": "object"
           }
         */
        /** Describing which chunk to use, instead of the default one, which would get chosen based on parameter ID and name. */
        chunk?: {
            /**
             * @swagger {components.schemas.CommonsParameterSType.properties.asset.properties.chunk.properties} id {
                 "description": "ID of the chunk to be used.",
                 "type": "string"
               }
             */
            /** ID of the chunk to be used. */
            id?: string;

            /**
             * @swagger {components.schemas.CommonsParameterSType.properties.asset.properties.chunk.properties} name {
                 "description": "Name attribute of the chunk to be used.",
                 "type": "string"
               }
             */
            /** Name attribute of the chunk to be used. */
            name?: string;
        };
    };
}
