// @shapediver/api.geometry-api-dto-v2/dist/commons/SdCommonModelComputation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShapeDiverCommonsModelComputationStatus = void 0;
var ShapeDiverCommonsModelComputationStatus;
(function (ShapeDiverCommonsModelComputationStatus) {
    ShapeDiverCommonsModelComputationStatus["SUCCESS"] = "success";
    ShapeDiverCommonsModelComputationStatus["TIMEOUT"] = "timeout";
    ShapeDiverCommonsModelComputationStatus["CHECK_CONFIRMED"] = "checkconfirmed";
    ShapeDiverCommonsModelComputationStatus["CHECK_DENIED"] = "checkdenied";
    ShapeDiverCommonsModelComputationStatus["CHECK_PENDING"] = "checkpending";
    ShapeDiverCommonsModelComputationStatus["MAX_COMBINED_ASSET_SIZE_EXCEEDED"] = "maxcombinedassetsizeexceeded";
    ShapeDiverCommonsModelComputationStatus["MAX_DB_SIZE_PER_OUTPUT_EXCEEDED"] = "maxdbsizeperoutputexceeded";
    ShapeDiverCommonsModelComputationStatus["MAX_PARTS_PER_OUTPUT_EXCEEDED"] = "maxpartsperoutputexceeded";
    ShapeDiverCommonsModelComputationStatus["MAX_ASSET_PARTS_PER_OUTPUT_EXCEEDED"] = "maxassetpartsperoutputexceeded";
    ShapeDiverCommonsModelComputationStatus["MAX_TRANSFORMATIONS_PER_OUTPUT_EXCEEDED"] = "maxtransformationsperoutputexceeded";
    ShapeDiverCommonsModelComputationStatus["MAX_PARTS_EXCEEDED"] = "maxpartsexceeded";
    ShapeDiverCommonsModelComputationStatus["MAX_ASSET_PARTS_EXCEEDED"] = "maxassetpartsexceeded";
    ShapeDiverCommonsModelComputationStatus["RECOVERABLE_ERROR"] = "recoverableerror";
    ShapeDiverCommonsModelComputationStatus["UNRECOVERABLE_ERROR"] = "unrecoverableerror";
    ShapeDiverCommonsModelComputationStatus["NO_OUTPUT_DATA_FOR_DEFAULT_PARAMETER_VALUES"] = "nooutputdatafordefaultparametervalues";
    ShapeDiverCommonsModelComputationStatus["MODEL_WITHOUT_GEOMETRY_OUTPUT"] = "modelwithoutgeometryoutput";
    ShapeDiverCommonsModelComputationStatus["UNKNOWN"] = "unknown";
})(ShapeDiverCommonsModelComputationStatus || (exports.ShapeDiverCommonsModelComputationStatus = ShapeDiverCommonsModelComputationStatus = {}));
