// @shapediver/viewer.settings/dist/versions/v6_1/Converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertToPrevious = exports.convertFromPrevious = void 0;
const convertFromPrevious = (s, v) => {
    const oldSettings = s;
    const settings = {
        settings_version: '6.1',
        ar: oldSettings.ar,
        build_date: oldSettings.build_date,
        build_version: oldSettings.build_version,
        camera: oldSettings.camera,
        general: oldSettings.general,
        light: oldSettings.light,
        session: oldSettings.session,
        environment: oldSettings.environment,
        environmentGeometry: {
            gridColor: oldSettings.environmentGeometry.gridColor,
            gridVisibility: oldSettings.environmentGeometry.gridVisibility,
            groundPlaneColor: oldSettings.environmentGeometry.groundPlaneColor,
            groundPlaneVisibility: oldSettings.environmentGeometry.groundPlaneVisibility,
            groundPlaneShadowColor: oldSettings.environmentGeometry.groundPlaneShadowColor,
            groundPlaneShadowVisibility: oldSettings.environmentGeometry.groundPlaneShadowVisibility,
            contactShadowVisibility: false,
            contactShadowOpacity: 1,
            contactShadowBlur: 1.5,
            contactShadowHeight: 0.05,
            contactShadowDarkness: 2.5
        },
        rendering: oldSettings.rendering,
        postprocessing: oldSettings.postprocessing,
        material: oldSettings.material ? oldSettings.material : {
            defaultMaterialColor: oldSettings.general.defaultMaterialColor,
            materialOverrideType: undefined
        }
    };
    /**
     * SETTINGS OBJECTS THAT DID CHANGE
     */
    return settings;
};
exports.convertFromPrevious = convertFromPrevious;
const convertToPrevious = (s, v) => {
    const newSettings = s;
    const settings = {
        settings_version: '6.0',
        ar: newSettings.ar,
        build_date: newSettings.build_date,
        build_version: newSettings.build_version,
        camera: newSettings.camera,
        general: newSettings.general,
        light: newSettings.light,
        session: newSettings.session,
        environment: newSettings.environment,
        environmentGeometry: {
            gridColor: newSettings.environmentGeometry.gridColor,
            gridVisibility: newSettings.environmentGeometry.gridVisibility,
            groundPlaneColor: newSettings.environmentGeometry.groundPlaneColor,
            groundPlaneVisibility: newSettings.environmentGeometry.groundPlaneVisibility,
            groundPlaneShadowColor: newSettings.environmentGeometry.groundPlaneShadowColor,
            groundPlaneShadowVisibility: newSettings.environmentGeometry.groundPlaneShadowVisibility
        },
        rendering: newSettings.rendering,
        postprocessing: newSettings.postprocessing,
        material: newSettings.material
    };
    /**
     * SETTINGS OBJECTS THAT DID CHANGE
     */
    return settings;
};
exports.convertToPrevious = convertToPrevious;
