// @shapediver/sdk.geometry-api-sdk-core/dist/ShapeDiverErrors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShapeDiverResponseError = exports.ShapeDiverRequestError = exports.ShapeDiverError = exports.ShapeDiverErrorType = void 0;
exports.ShapeDiverErrorType = {
    Generic: "sd_gb_sdk_generic",
    Request: "sd_gb_sdk_request",
    Response: "sd_gb_sdk_response",
};
class ShapeDiverError extends Error {
    constructor(message) {
        super(message);
        this.errorType = exports.ShapeDiverErrorType.Generic;
    }
}
exports.ShapeDiverError = ShapeDiverError;
class ShapeDiverRequestError extends ShapeDiverError {
    constructor(desc, message) {
        super(message);
        this.desc = desc;
        this.errorType = exports.ShapeDiverErrorType.Request;
    }
}
exports.ShapeDiverRequestError = ShapeDiverRequestError;
class ShapeDiverResponseError extends ShapeDiverError {
    constructor(message, status, error, desc, headers) {
        super(message);
        this.status = status;
        this.error = error;
        this.desc = desc;
        this.headers = headers;
        this.errorType = exports.ShapeDiverErrorType.Response;
    }
}
exports.ShapeDiverResponseError = ShapeDiverResponseError;
